src/Deinterleave.cpp

```diff
@@ -661,15 +661,137 @@ class Deinterleaver : public IRGraphMutator {
         return Cast::make(t, mutate(op->value));
     }
 
-    // FIXME: Implement me.
     Expr visit(const Reinterpret *op) override {
-        if (op->type.is_scalar()) {
-            return op;
-        } else {
-            // FIXME: this is broken.
-            Type t = op->type.with_lanes(lanes.size());
+        Type src_ty = op->value.type();
+        int src_lanes = src_ty.lanes();
+        Type tgt_ty = op->type;
+        int tgt_lanes = tgt_ty.lanes();
+
+        if (src_lanes == tgt_lanes) {
+            // If the number of lanes wasn't changing, just recurse further.
+            Type t = tgt_ty.with_lanes(lanes.size());
             return Reinterpret::make(t, mutate(op->value));
         }
+
+        // Otherwise, since `Reinterpret` never changes the total bit count,
+        // but the lane count change, so must the width of the lanes, and to
+        // simplify further logic, it is best to do so with 'uint'-typed lanes.
+        if (!src_ty.is_uint() || !tgt_ty.is_uint()) {
+            auto reinterpret = [](Expr e, Type t, halide_type_code_t code) {
+                t = t.with_code(code);
+                if (e.type() != t) {
+                    e = Reinterpret::make(t, e);
+                }
+                return e;
+            };
+
+            Expr in = op->value;
+            // Prepare - ensure that the type is `uint`.
+            in = reinterpret(in, in.type(), Type::UInt);
+            // Perform the actual lane splitting/merging, in `uint`.
+            in = reinterpret(in, tgt_ty, Type::UInt);
+            // Recover - convert to the expected output type.
+            in = reinterpret(in, tgt_ty, tgt_ty.code());
+            return mutate(in);
+        }
+
+        auto [min_lanes, max_lanes] = std::minmax(src_lanes, tgt_lanes);
+
+        // Likewise, we really need an integral scaling factor.
+        // If it is not, we, again, want to expand.
+        if (max_lanes % min_lanes != 0) {
+            // We want to avoid the precision loss when back-transforming the
+            // demanded bits @ tgt_ty into demanded bits @ src_ty, so we want
+            // such an immediate type, that has N times more lanes than tgt_ty,
+            // while *also* having M times more lanes than src_ty,
+            // while minimizing N and M.
+            int lanes_lcm = std::lcm(src_lanes, tgt_lanes);  // Optimal.
+            internal_assert(lanes_lcm % tgt_ty.lanes() == 0);
+            int lane_prescale = lanes_lcm / tgt_ty.lanes();
+            internal_assert(tgt_ty.bits() % lane_prescale == 0);
```

Comment:
I have brute-force checked this for every vector length in 1:4096 for every possible src/tgt lane count.